### PR TITLE
Databricks/client creds input validation

### PR DESCRIPTION
**Pull request by ihkap11**
# Description
Add validators for Databricks credentials - token id and cluster id, to ensure  proper format to avoid confusion with users.

Tests failing before: 
`===== 104 failed, 383 passed, 14 xfailed, 227 warnings, 5 errors in 45.88s =====`
Tests failing after: 
`===== 103 failed, 384 passed, 19 xfailed, 227 warnings, 5 errors in 45.81s =====`

Issue: https://github.com/featureform/featureform/issues/1143

### Does this correspond to an open issue?
[](https://github.com/featureform/featureform/issues/1143)

### Select type(s) of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

# Checklist:

- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have fixed any merge conflicts

### CODE REVIEW

File: client/src/featureform/resources.py [hunk 1]
```diff
@@ -2278,9 +2280,27 @@ def __init__(
                 "The DatabricksCredentials requires only one credentials set ('username' and 'password' or 'host' and 'token' set.)"
             )
 
-        if not cluster_id:
+        if not self.cluster_id:
             raise Exception("Cluster_id of existing cluster must be provided")
 
+        if not self._validate_cluster_id():
+            raise ValueError(
+                "Cluster_id must follow the format: xxx-xxxxxx-xxxxxxxx, where x is alphanumeric"
```
Comment (sdreyer) [inline client/src/featureform/resources.py:2288]: Lets change this to something like:
`f"Invalid cluster_id: expected id in the format 'xxx-xxxxxx-xxxxxxxx' but received '{self.cluster_id}'`

File: client/src/featureform/resources.py [hunk 2]
```diff
@@ -2278,9 +2280,27 @@ def __init__(
...
+            )
+
+        if self.host and not self._validate_token():
+            raise ValueError(
+                "Token must start with 'dapi' followed by alphanumeric eg ex: dapixxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx-x"
```
Comment (sdreyer) [inline client/src/featureform/resources.py:2293]: Similarly
`f"Invalid token: expected token in the format 'dapixxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx-x' but received '{self.token}'`

File: client/src/featureform/resources.py [hunk 3]
```diff
@@ -2278,9 +2280,27 @@ def __init__(
...
+            )
+
+    def _validate_cluster_id(self):
+        cluster_id_regex = r"^\w{3}-\w{6}-\w{8}$"
+        return re.match(cluster_id_regex, self.cluster_id)
+
+    def _validate_token(self):
+        token_regex = r"^dapi[\w-]+$"
```
Comment (sdreyer) [inline client/src/featureform/resources.py:2301]: Should limit the number of characters. I believe the format I posted has the correct # of characters overall

**Review by sdreyer — COMMENTED** [review]
Looking good! A few changes

File: client/src/featureform/resources.py (hunk 1 above)
Comment (ihkap11) [reply client/src/featureform/resources.py]: Changed here - 246fc8d0d2b2949838806b0b795b04f69daba214

File: client/src/featureform/resources.py (hunk 2 above)
Comment (ihkap11) [reply client/src/featureform/resources.py]: Changed here - 246fc8d0d2b2949838806b0b795b04f69daba214

File: client/src/featureform/resources.py (hunk 3 above)
Comment (ihkap11) [reply client/src/featureform/resources.py]: Got it! Just need a little clarification on the following:

Databricks example [here](https://www.databricks.com/blog/2022/08/03/how-to-build-a-marketing-analytics-solution-using-fivetran-and-dbt-on-the-databricks-lakehouse.html#:~:text=token%20(dapiXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX)%3A) uses the following token format:
`dapiXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX`
Our's has a slight variation:
`dapixxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx-x`
If the placement of hyphen always remains the same, here is the fix: a43176c4de6645b577761bc7f7986c8fc87faf99
If not, let me know.

**Review by sdreyer — APPROVED** [review]

File: client/src/featureform/resources.py (hunk 3 above)
Comment (sdreyer) [reply client/src/featureform/resources.py]: Yeah its interesting. Ours always have the dash in it

Comment (sdreyer) [reply client/src/featureform/resources.py]: Lets leave for now, we're adding some updates around this so we may add more variations later

**Pull request merged**